### PR TITLE
Fixed large pcolormesh rendering with bbox_inches='tight'.

### LINKED ISSUE #3095: Memory issue when plotting large arrays with pcolormesh
Plotting large-ish arrays using pcolormesh with current (e2c5918) matplotlib master looks like it might have a memory leak. Running the following code on matplotlib v1.3.1 produces a `QuadMesh` in approximately a second. However running the same code on matplotlib master floods the memory until either the pcolormesh call is aborted or it runs out of memory.

```python
import matplotlib.pyplot as plt
import numpy as np

data = np.random.rand(3000*3000).reshape(3000, 3000)
plt.pcolormesh(data)
plt.show()
```

### PR DESCRIPTION
**Pull request by pelson**
Replaces #4017.
Closes #3095.

### CODE REVIEW

**Comment by tacaswell** [comment]
This does not account for having a non-trivial `offset` array.

**Comment by pelson** [comment]
> This does not account for having a non-trivial offset array.

Can you provide a simple example? I've never properly understood the offsets in some collection contexts.

**Comment by tacaswell** [comment]
```python
import numpy as np
import matplotlib.pyplot as plt

fig, ax = plt.subplots(1, 1)

X, Y = np.meshgrid(range(25), range(20))

C = np.random.rand(*X.shape)
offsets = (np.random.randn(2, 24*19) 

ax.pcolormesh(X, Y, C, offsets=offsets, transOffset=ax.transData)

```
so

which has an odd extra shift from (0, 0) to (3, 2.3) for the lower left.  I am also not _really_ clear on why this is useful, but it is something that the code does.

**Comment by tacaswell** [comment]
The code for this in agg ends up in c++ layer and gets really hairy really fast.

**Comment by tacaswell** [comment]
```python
import numpy as np
import matplotlib.pyplot as plt

fig, ax = plt.subplots(1, 1)

X, Y = np.meshgrid(range(25), range(20))

C = np.random.rand(*X.shape)
offsets = np.random.randn(24*19, 2) * .1

ax.pcolormesh(X, Y, C)#, offsets=offsets, transOffset=ax.transData)
fig.savefig('so.png', bbox_inches='tight')
```

gives me

```
<ipython-input-26-382f89238547> in <module>()
----> 1 exec(compile(open('/tmp/ipython-5366aRU.py').read(), '/tmp/ipython-5366aRU.py', 'exec')) # PYTHON-MODE

/tmp/ipython-5366aRU.py in <module>()

/home/tcaswell/.virtualenvs/mpl_master_3k/lib/python2.7/site-packages/matplotlib-1.4.x.dev0-py2.7-linux-x86_64.egg/matplotlib/figure.pyc in savefig(self, *args, **kwargs)
   1474             self.set_frameon(frameon)
   1475 
-> 1476         self.canvas.print_figure(*args, **kwargs)
   1477 
   1478         if frameon:

/home/tcaswell/.virtualenvs/mpl_master_3k/lib/python2.7/site-packages/matplotlib-1.4.x.dev0-py2.7-linux-x86_64.egg/matplotlib/backends/backend_qt5agg.pyc in print_figure(self, *args, **kwargs)
    159 
    160     def print_figure(self, *args, **kwargs):
--> 161         FigureCanvasAgg.print_figure(self, *args, **kwargs)
    162         self.draw()
    163 

/home/tcaswell/.virtualenvs/mpl_master_3k/lib/python2.7/site-packages/matplotlib-1.4.x.dev0-py2.7-linux-x86_64.egg/matplotlib/backend_bases.pyc in print_figure(self, filename, dpi, facecolor, edgecolor, orientation, format, **kwargs)
   2166                 bbox_filtered = []
   2167                 for a in bbox_artists:
-> 2168                     bbox = a.get_window_extent(renderer)
   2169                     if a.get_clip_on():
   2170                         clip_box = a.get_clip_box()

/home/tcaswell/.virtualenvs/mpl_master_3k/lib/python2.7/site-packages/matplotlib-1.4.x.dev0-py2.7-linux-x86_64.egg/matplotlib/collections.pyc in get_window_extent(self, renderer)
    208         # TODO:check to ensure that this does not fail for
    209         # cases other than scatter plot legend
--> 210         return self.get_datalim(transforms.IdentityTransform())
    211 
    212     def _prepare_points(self):

/home/tcaswell/.virtualenvs/mpl_master_3k/lib/python2.7/site-packages/matplotlib-1.4.x.dev0-py2.7-linux-x86_64.egg/matplotlib/collections.pyc in get_datalim(self, transData)
   1707 
   1708     def get_datalim(self, transData):
-> 1709         return (self.get_transform() - transData).transform_bbox(self._bbox)
   1710 
   1711     @staticmethod

/home/tcaswell/.virtualenvs/mpl_master_3k/lib/python2.7/site-packages/matplotlib-1.4.x.dev0-py2.7-linux-x86_64.egg/matplotlib/transforms.pyc in transform_bbox(self, bbox)
   1352         requirement for matplotlib figures), see :class:`TransformedBbox`.
   1353         """
-> 1354         return Bbox(self._transform.transform(bbox.get_points()))
   1355 
   1356     def get_affine(self):

```

File: lib/matplotlib/transforms.py [hunk 1]
```diff
@@ -1344,6 +1344,15 @@ def transform_non_affine(self, values):
...
         return values
 
+    def transform_bbox(self, bbox):
+        """
+        Transform the given bounding box.
+
+        Note, for smarter transforms including caching (a common
+        requirement for matplotlib figures), see :class:`TransformedBbox`.
+        """
+        return Bbox(self._transform.transform(bbox.get_points()))
```
Comment (tacaswell) [inline lib/matplotlib/transforms.py]: I think this line should be `return Bbox(self..transform(bbox.get_points()))`

Comment (pelson) [inline lib/matplotlib/transforms.py]: Agreed.

**Comment by pelson** [comment]
Thanks for the example @tacaswell. I did not know you could do that!
I'll take a look at dealing with the offsets, but ultimately if this is holding up a release, I'd be comfortable trading a major performance bug with pcolormesh, for a minor cut-off known bug with a not-well-used part of the same routine.

**Comment by tacaswell** [comment]
I concur about the trade off.

**Pull request merged**